include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.cpp

```diff
@@ -0,0 +1,334 @@
+
+  // loop over all boundary faces
+  for(unsigned int face = face_range.first; face < face_range.second; face++)
+  {
+    integrator.reinit(face);
+    integrator.read_dof_values(src);
+
+    // we only need the gradient
+    integrator.evaluate(dealii::EvaluationFlags::gradients);
+
+    // loop over all quadrature points
+    for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+    {
+      // calculate velocity gradient
+      tensor velocity_gradient = integrator.get_gradient(q);
+
+      // get the coefficients
+      scalar viscosity = Base::viscous_kernel->get_coefficient_face(face, q);
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+
+      // set the coefficients
+      Base::viscous_kernel->set_coefficient_face(face, q, viscosity);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::add_generalized_newtonian_viscosity(
+  scalar &       viscosity,
+  tensor const & velocity_gradient) const
+{
+  tensor symmetric_velocity_gradient = 0.5 * (velocity_gradient + transpose(velocity_gradient));
+
+  scalar shear_rate =
+    std::sqrt(2.0 * scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient));
+
+  scalar viscosity_factor;
+  compute_viscosity_factor(viscosity_factor, shear_rate);
+
+  viscosity += viscosity_factor * viscosity_margin;
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::compute_viscosity_factor(scalar &       viscosity_factor,
+                                                                 scalar const & shear_rate) const
+{
+  switch(generalized_newtonian_model)
+  {
+    case GeneralizedNewtonianViscosityModel::Undefined:
+      AssertThrow(generalized_newtonian_model != GeneralizedNewtonianViscosityModel::Undefined,
+                  dealii::ExcMessage("parameter must be defined"));
+      break;
+    case GeneralizedNewtonianViscosityModel::GeneralizedCarreauYasuda:
+      generalized_carreau_yasuda_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Carreau:
+      carreau_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Cross:
+      cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::SimplifiedCross:
+      simplified_cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::PowerLaw:
+      power_law_model(viscosity_factor, shear_rate);
+      break;
+    default:
+      AssertThrow(
+        false, dealii::ExcMessage("This GeneralizedNewtonianViscosityModel is not implemented."));
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::generalized_carreau_yasuda_model(
+  scalar &       viscosity_factor,
+  scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
```

Comment:
I think this comment should be given in the function `compute_viscosity_factor()`, explaining which part of the formula is computed by function (as far as I can see it, only the last part with []^[]). This would improve my understanding of the code. Then, duplicating the code here and below is superfluous.